src/debug/validatePreDebug.ts

```diff
@@ -4,34 +4,74 @@
  *--------------------------------------------------------------------------------------------*/
 
 import * as azureStorage from "azure-storage";
+import * as fse from 'fs-extra';
+import * as path from 'path';
 import * as vscode from 'vscode';
-import { parseError } from "vscode-azureextensionui";
+import { AzureWizard, IActionContext, parseError } from "vscode-azureextensionui";
+import { getWorkspaceSetting } from "../../extension.bundle";
+import { AzureWebJobsStorageExecuteStep } from "../commands/appSettings/AzureWebJobsStorageExecuteStep";
+import { AzureWebJobsStoragePromptStep } from "../commands/appSettings/AzureWebJobsStoragePromptStep";
+import { IAzureWebJobsStorageWizardContext } from "../commands/appSettings/IAzureWebJobsStorageWizardContext";
 import { tryGetFunctionProjectRoot } from '../commands/createNewProject/verifyIsProject';
-import { isWindows, localEmulatorConnectionString, localSettingsFileName } from "../constants";
+import { functionJsonFileName, isWindows, localEmulatorConnectionString, localSettingsFileName, projectLanguageSetting } from "../constants";
 import { ext } from "../extensionVariables";
+import { ParsedFunctionJson } from "../funcConfig/function";
 import { azureWebJobsStorageKey, getAzureWebJobsStorage } from "../funcConfig/local.settings";
 import { validateFuncCoreToolsInstalled } from '../funcCoreTools/validateFuncCoreToolsInstalled';
 import { localize } from '../localize';
+import { getFunctionFolders } from "../tree/localProject/LocalFunctionsTreeItem";
+import { supportsLocalProjectTree } from "../tree/localProject/supportsLocalProjectTree";
 import { getDebugConfigs, isDebugConfigEqual } from '../vsCodeConfig/launch';
 
 export interface IPreDebugValidateResult {
     workspace: vscode.WorkspaceFolder;
     shouldContinue: boolean;
 }
 
-export async function preDebugValidate(debugConfig: vscode.DebugConfiguration): Promise<IPreDebugValidateResult> {
+export async function preDebugValidate(context: IActionContext, debugConfig: vscode.DebugConfiguration): Promise<IPreDebugValidateResult> {
     const workspace: vscode.WorkspaceFolder = getMatchingWorkspace(debugConfig);
-    let shouldContinue: boolean = await validateFuncCoreToolsInstalled();
-    if (shouldContinue) {
-        const projectPath: string | undefined = await tryGetFunctionProjectRoot(workspace.uri.fsPath, true /* suppressPrompt */);
-        if (projectPath) {
-            shouldContinue = await validateEmulatorIsRunning(projectPath);
+    let shouldContinue: boolean;
+    context.telemetry.properties.debugType = debugConfig.type;
+
+    try {
+        context.telemetry.properties.lastValidateStep = 'funcInstalled';
+        shouldContinue = await validateFuncCoreToolsInstalled();
+
+        if (shouldContinue) {
```

Comment:
I don't really understand why you have to return this value.  Either way, it seems like you exit the command (though for one instance, you return undefined).  But couldn't you just exit out of the function?

Reply:
Discussed offline